c++/src/kj/async-coroutine-test.c++

```diff
@@ -0,0 +1,183 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <kj/async.h>
+#include <kj/compat/http.h>
+#include <kj/debug.h>
+#include <kj/test.h>
+
+namespace kj {
+namespace {
+
+#ifdef KJ_HAS_COROUTINE
+
+template <typename T>
+Promise<kj::Decay<T>> identity(T&& value) {
+  co_return kj::fwd<T>(value);
+}
+// Work around a bonkers MSVC ICE with a separate overload.
+Promise<const char*> identity(const char* value) {
+  co_return value;
+}
+
+KJ_TEST("Identity coroutine") {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  KJ_EXPECT(identity(123).wait(waitScope) == 123);
+  KJ_EXPECT(*identity(kj::heap(456)).wait(waitScope) == 456);
+
+  {
+    auto p = identity("we can cancel the coroutine");
+  }
+}
+
+template <typename T>
+Promise<T> simpleCoroutine(kj::Promise<T> result, kj::Promise<bool> dontThrow = true) {
+  KJ_ASSERT(co_await dontThrow);
+  co_return co_await result;
+}
+
+KJ_TEST("Simple coroutine test") {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  simpleCoroutine(kj::Promise<void>(kj::READY_NOW)).wait(waitScope);
+
+  KJ_EXPECT(simpleCoroutine(kj::Promise<int>(123)).wait(waitScope) == 123);
+}
+
+KJ_TEST("Exceptions propagate through layered coroutines") {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  // This exercises the Coroutine<T>::unhandled_exception() path.
+  auto throwy = simpleCoroutine(kj::Promise<int>(kj::NEVER_DONE), false);
+
+  // And this exercises the exception short-circuit in Adapter::fire().
+  KJ_EXPECT_THROW_RECOVERABLE(FAILED, simpleCoroutine(kj::mv(throwy)).wait(waitScope));
+}
+
+KJ_TEST("Exceptions before the first co_await don't leak, but reject the promise") {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  auto throwEarly = []() -> Promise<void> {
+    KJ_FAIL_ASSERT("test exception");
+    co_return;
+  };
+
+  auto throwy = throwEarly();
+
+  KJ_EXPECT_THROW_RECOVERABLE(FAILED, throwy.wait(waitScope));
+}
+
+KJ_TEST("Coroutines can be canceled while suspended") {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  size_t count = 0;
+
+  struct Counter {
+    size_t& count;
+    Counter(size_t& count): count(count) {}
+    ~Counter() { ++count; }
+    KJ_DISALLOW_COPY(Counter);
+  };
+
+  {
+    auto neverDone = kj::Promise<int>(kj::NEVER_DONE);
+    neverDone = neverDone.attach(kj::heap<Counter>(count));
```

Comment:
This verifies that the `neverDone` promise is itself canceled. Great!

It'd also be nice to verify that local variables declared in the coroutine are destroyed on cancellation.

Reply:
Test here: https://github.com/capnproto/capnproto/pull/1104/commits/9407ece3275c5fafb2561f62283155e0484bcd07